### PR TITLE
ROX-15164: Reconciliation loop for declarative configuration

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR adds the reconciliation loop for declarative configurations.

After the contents of the declarative configuration mounts will be watched by multiple watch handlers (introduced within [ROX-14147](https://issues.redhat.com/browse/ROX-14147), the next step is to start with the reconciliation of the transformed proto messages.

This includes the following:
- Creating a "complete" view of all transformed proto messages, keyed by their respective watch handler directory.
- Creating a loop that will continously iterate over the transformed proto messages and allow for upserting proto messages as well as deleting proto messages.

Besides these two points, the following was done as well:
- Instead of specifying the watch handler intervals as well as the reconciliation loop interval as `const`s, exposed environment variables for these two settings for further customization if required in some environments.
- The reconciliation loop includes a short-circuit functionality, bypassing the reconciliation loop interval to trigger the reconciliation when changes are reported by the watch handlers.

As follow-ups, the following will be required:
- Handle the upsertion of proto messages within the reconciliation step, covered by [ROX-14693](https://issues.redhat.com/browse/ROX-14693).
- Handle the deletion of proto messages within the reconciliation step, covered by [ROX-14694](https://issues.redhat.com/browse/ROX-14694).

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why
you did not do so. Valid reasons include, for example, "CI is sufficient",
"No testable changes". Feel free to attach JSON snippets, curl commands,
screenshots.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: central/declarativeconfig/manager_impl.go [hunk 1]
```diff
@@ -8,39 +8,51 @@ import (
...
 // New creates a new instance of Manager.
 // Note that it will not watch the declarative configuration directories when created, only after
-// WatchDeclarativeConfigDir has been called.
+// ReconcileDeclarativeConfigurations has been called.
 func New() Manager {
 	return &managerImpl{
-		universalTransformer: transform.New(),
+		universalTransformer:         transform.New(),
+		transformedMessagesByHandler: map[string]protoMessagesByType{},
+		reconciliationTickerDuration: env.DeclarativeConfigReconcileInterval.DurationSetting(),
```
Comment (ivan-degtiarenko) [inline central/declarativeconfig/manager_impl.go:50]: I would expect these parameters to be passed to the `New` method and extracted from `env` in a Singleton

File: central/declarativeconfig/manager_impl.go [hunk 2]
```diff
@@ -87,5 +109,64 @@ func (m *managerImpl) ReconcileDeclarativeConfigs(contents [][]byte) {
 			transformedConfigurations[protoType] = append(transformedConfigurations[protoType], protoMessages...)
 		}
 	}
-	// No-op, see the TODOs within the function comment.
+
+	m.transformedMessagesMutex.Lock()
+	defer m.transformedMessagesMutex.Unlock()
```
Comment (ivan-degtiarenko) [inline central/declarativeconfig/manager_impl.go:114]: there's no reason to hold the lock while executing `m.shortCircuitReconciliationLoop()`

File: central/declarativeconfig/manager_impl.go [hunk 3]
```diff
@@ -87,5 +109,64 @@ func (m *managerImpl) ReconcileDeclarativeConfigs(contents [][]byte) {
...
+			m.shortCircuitSignal.Reset()
+			m.runReconciliation()
+		case <-m.reconciliationTicker.C:
+			m.runReconciliation()
+		}
+	}
+}
+
+func (m *managerImpl) runReconciliation() {
+	// We shouldn't trigger a parallel reconciliation run, hence we should ensure that the flag is not set to true.
```
Comment (ivan-degtiarenko) [inline central/declarativeconfig/manager_impl.go:150]: I just now realised that we can't trigger parallel runs either way - all `runReconciliation` calls are happening in the one goroutine. Perhaps I'm missing something?

File: central/declarativeconfig/manager_impl.go [hunk 4]
```diff
@@ -87,5 +109,64 @@ func (m *managerImpl) ReconcileDeclarativeConfigs(contents [][]byte) {
...
+}
+
+func (m *managerImpl) startReconciliationLoop() {
+	m.reconciliationTicker = time.NewTicker(m.reconciliationTickerDuration)
+
+	go m.reconciliationLoop()
+}
+
+func (m *managerImpl) reconciliationLoop() {
+	// While we currently do not have an exist in the form of "stopping" the reconciliation, still, ensure that
```
Comment (ivan-degtiarenko) [inline central/declarativeconfig/manager_impl.go:135]:
```suggestion
	// While we currently do not have an exit in the form of "stopping" the reconciliation, still, ensure that
```

File: central/declarativeconfig/manager_impl.go [hunk 5]
```diff
@@ -87,5 +109,64 @@ func (m *managerImpl) ReconcileDeclarativeConfigs(contents [][]byte) {
...
+	m.transformedMessagesMutex.Lock()
+	defer m.transformedMessagesMutex.Unlock()
+	m.transformedMessagesByHandler[handlerID] = transformedConfigurations
+	m.shortCircuitReconciliationLoop()
+}
+
+// shortCircuitReconciliationLoop will short circuit the reconciliation loop and trigger a reconciliation loop run.
+// Note that the reconciliation loop will not be run if:
+//   - the short circuit loop signal has not been reset yet and is de-duped.
+//   - a current reconciliation loop run is in progress.
```
Comment (ivan-degtiarenko) [inline central/declarativeconfig/manager_impl.go:122]: I don't think that's the case - my understanding is that `for select` is a single-thread, consecutive execution mechanism. Thus after the reconciliation is finished, we go into a new iteration of the cycle and handle short circuit signal

Comment (dhaus67) [reply central/declarativeconfig/manager_impl.go:122]: Addressed within the comment above.

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM 👍

**Pull request merged**